### ISSUE TITLE
Discussing how CV validation metrics are aggregated.

### ISSUE THREAD

**Issue by samFarrellDay** [microsoft/LightGBM#5161]
Sorry if this has been discussed before, I couldn't find an issue about this.

## Summary

As it is, CV results are aggregated by taking the average of the individual loss metrics on each validation set. This works fine for some loss metrics (logloss, RMSE) but in some cases it does not work well for others (AUC, maybe others?). For example, if you are performing leave-one-out cross validation, it is impossible to calculate the AUC because there is only 1 sample in the validation set, and the AUC cannot be calculated on one sample. In fact, lightgbm throws a warning if the number of splits is greater than the least populated class.

I think a good, easy solution would be to aggregate the cv loss metrics by applying the loss metric on the entire cv prediction array at once. This might even improve performance, since many loss metric functions are meant to take advantage of vectorization, and you would only be calling the loss function once on the entire cv prediction array.

Pseudo-code for the current implementation:
```mean(loss(v_pred, v_target) for v in valid_sets)```

Pseudo-code the the proposed implementation:
```loss(cv_preds, target)```

For some applications, these would be theoretically equivalent. See the following code, where the calculated cv loss from lightgbm is the same as the out of fold predictions:

```
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import log_loss
import numpy as np
import lightgbm as lgb

# Make data
skf = model_selection.StratifiedKFold(n_splits=10, shuffle=True, random_state=1991)
target = np.random.randint(2,size=100)
dat = np.random.randint(100, size=(100,5))
dtrain = lgb.Dataset(dat, target)

# cv
lgbcv = lgb.cv(params={"objective": "binary", "verbose":-1}, train_set=dtrain, folds=skf, return_cvbooster=True)

# Get the cross-validated predictions
skf_folds = skf.split(dat, target)
cv_prediction_array = np.empty(100)
for n, (t, v) in enumerate(skf_folds):
    cv_prediction_array[v] = lgbcv["cvbooster"].boosters[n].predict(dat[v,:])

# Logloss from cv predictions is the same as the average validation logloss.
print(log_loss(target, cv_prediction_array) / lgbcv['binary_logloss-mean'][-1])
```

### Cons:
You could no longer calculate the standard deviation of the validation losses. I'm not well versed on how this would affect some of the more esoteric objectives, like learning-to-rank, and I don't know if theoretically the effective re-weighting of samples would constitute data leakage.
